source/kraken/fill_disruption_from_chaos.cpp

```diff
@@ -523,148 +831,205 @@ struct delete_impacts_visitor : public apply_impacts_visitor {
+        // it should be removed from jp
+        std::vector<nt::JourneyPatternKey> jp_to_remove_from_pool;
+        boost::for_each(pt_data.journey_patterns_pool,
+                [&](const decltype(pt_data.journey_patterns_pool)::value_type& key_pair){
+            if(! contains(pt_data.journey_patterns_map, key_pair.second->uri)) {
+                jp_to_remove_from_pool.push_back(key_pair.first);
+            }
+        });
+        boost::for_each(jp_to_remove_from_pool, [&](const nt::JourneyPatternKey& key){
+            pt_data.journey_patterns_pool.erase(key);
+        });
+    }
 
+    template<typename T>
+    void reassign_validity_pattern(const std::vector<T*>& stop_points) {
         LOG4CPLUS_DEBUG(log4cplus::Logger::getInstance("log"),
-                        "Reassigning validity_pattern");
-        for (auto stop_point : stop_area->stop_point_list) {
-            for (auto journey_pattern_point : stop_point->journey_pattern_point_list) {
-                apply_impacts_visitor::operator ()(journey_pattern_point->journey_pattern);
+                "Reassigning validity_pattern");
+        for (const auto* stop_point : stop_points) {
+            for (const auto* journey_pattern_point : stop_point->journey_pattern_point_list) {
+                apply_impacts_visitor::operator()(journey_pattern_point->journey_pattern);
             }
         }
-
         LOG4CPLUS_DEBUG(log4cplus::Logger::getInstance("log"),
-                        nb_vj_reassigned << " validity pattern reassigned");
+                nb_vj_reassigned << " validity pattern reassigned");
+    }
+
+
+    void operator()(const nt::StopPoint* stop_point) override {
+
+        this->log_start_action(stop_point->uri);
+
+        delete_impacts_impl();
+
+        reassign_validity_pattern(std::vector<const nt::StopPoint*>{stop_point});
+
+        this->log_end_action(stop_point->uri);
+    }
+
+    void operator()(const nt::StopArea* stop_area) override{
+        this->log_start_action(stop_area->uri);
+
+        delete_impacts_impl();
+
+        reassign_validity_pattern(stop_area->stop_point_list);
+
         this->log_end_action(stop_area->uri);
     }
 };
 
 void delete_impact(boost::shared_ptr<nt::new_disruption::Impact>impact,
-                          nt::PT_Data& pt_data, const nt::MetaData& meta) {
+        nt::PT_Data& pt_data, const nt::MetaData& meta) {
     if (impact->severity->effect != nt::new_disruption::Effect::NO_SERVICE) {
         return;
     }
     LOG4CPLUS_DEBUG(log4cplus::Logger::getInstance("log"),
-                    "Deleting impact: " << impact.get()->uri);
+            "Deleting impact: " << impact.get()->uri);
     delete_impacts_visitor v(impact, pt_data, meta);
     boost::for_each(impact->informed_entities, boost::apply_visitor(v));
+
+    // We should re-apply all other impacts
+    boost::for_each(v.other_impacts, [&](const decltype(std::begin(v.other_impacts))::value_type& i){
```

Comment:
collections declare `const_reference`, better that constructing it from `value_type`.